src/server/pkg/hashtree/cache.go

```diff
@@ -4,35 +4,81 @@ import (
 	"fmt"
 	"io"
 	"reflect"
+	"sync"
 
-	lru "github.com/hashicorp/golang-lru"
+	"github.com/hashicorp/golang-lru/simplelru"
 	"github.com/pachyderm/pachyderm/src/server/pkg/localcache"
 	"github.com/sirupsen/logrus"
 )
 
 // Cache is an LRU cache for hashtrees.
 type Cache struct {
-	*lru.Cache
+	lruCache  simplelru.LRUCache
```

Comment:
Rather than using `lru.Cache`, we now use its lower-level `simplelru.LRUCache` which does not implement any synchronization locks, so we can define our own.  In addition, we only ever used one set of operations for the cache (get or add by a specific key), so that was rolled into one atomic function (GetOrAdd).  See the comment on that function below for more details.